clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -1040,84 +1124,156 @@ protected void produceToStoreBufferServiceOrKafka(
       }
-      totalBytesRead += record.getPayloadSize();
-      // Update the latest message consumed time
+
+      hostLevelIngestionStats.recordStorageQuotaUsed(storageUtilizationManager.getDiskQuotaUsage());
+    }
+  }
+
+  protected void produceToStoreBufferServiceOrKafkaInBatch(
+      Iterable<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records,
+      PubSubTopicPartition topicPartition,
+      PartitionConsumptionState partitionConsumptionState,
+      String kafkaUrl,
+      int kafkaClusterId) throws InterruptedException {
+    long totalBytesRead = 0;
+    ValueHolder<Double> elapsedTimeForPuttingIntoQueue = new ValueHolder<>(0d);
+    boolean metricsEnabled = emitMetrics.get();
+    long beforeProcessingBatchRecordsTimestampMs = System.currentTimeMillis();
+    /**
+     * Split the records into mini batches.
+     */
+    int batchSize = serverConfig.getAaWCWorkloadParallelProcessingThreadPoolSize();
+    List<List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>>> batches = new ArrayList<>();
+    List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> ongoingBatch = new ArrayList<>(batchSize);
+    Iterator<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> iter = records.iterator();
+    while (iter.hasNext()) {
+      PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> record = iter.next();
       if (partitionConsumptionState != null) {
-        partitionConsumptionState.setLatestMessageConsumedTimestampInMs(beforeProcessingBatchRecordsTimestampMs);
+        partitionConsumptionState.setLatestPolledMessageTimestampInMs(beforeProcessingBatchRecordsTimestampMs);
+      }
+      if (!shouldProcessRecord(record)) {
+        if (partitionConsumptionState != null) {
+          partitionConsumptionState.updateLatestIgnoredUpstreamRTOffset(kafkaUrl, record.getOffset());
+        }
+        continue;
+      }
+      waitReadyToProcessRecord(record);
+      ongoingBatch.add(record);
+      if (ongoingBatch.size() == batchSize) {
+        batches.add(ongoingBatch);
+        ongoingBatch = new ArrayList<>(batchSize);
+      }
+    }
+    if (!ongoingBatch.isEmpty()) {
+      batches.add(ongoingBatch);
+    }
+    if (batches.isEmpty()) {
+      return;
+    }
+    IngestionBatchProcessor ingestionBatchProcessor = getIngestionBatchProcessor();
+    if (ingestionBatchProcessor == null) {
+      throw new VeniceException(
+          "IngestionBatchProcessor object should present for store version: " + kafkaVersionTopic);
+    }
+    /**
+     * Process records batch by batch.
+     */
+    for (List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> batch: batches) {
+      List<ReentrantLock> locks = ingestionBatchProcessor.lockKeys(batch);
+      try {
+        long beforeProcessingPerRecordTimestampNs = System.nanoTime();
+        List<PubSubMessageProcessedResultWrapper<KafkaKey, KafkaMessageEnvelope, Long>> processedResults =
+            ingestionBatchProcessor.process(
```

Comment:
Please bear me with some dumb question - even though we synced offline but I am still a bit confused about the processing logic here: if there is control message such as SOS/EOS (which carries something like HB) how does batch processes process these messages? I checked the processFunction passed in, looks like it only assumes data message?

Reply:
Batch processing function will skip all the Control messages.